containerd-shim-spin-config.md

```diff
@@ -0,0 +1,372 @@
+    env:
+    - name: SPIN_HTTP_LISTEN_ADDR
+      value: "0.0.0.0:3000"
+    ...
+```
+
+The `containerd-shim-spin` instance will look for execution environment
+variables and configure execution accordingly.
+
+At first, this feels strange, since it differs from user's previous
+understanding of Linux container environment variables. Deploying a Spin app
+shouldn't be compared to deploying a Linux container. A Linux container has one
+virtualized environment and listener. In a Linux container context, environment
+variables are set in the app execution and affect the business logic. On the
+other hand, a Spin app can have N many triggers and components, each of which
+has its own virtualized environment. While Spin does support setting environment
+variables in components, they are static. We have never encouraged the use of
+`spin up --env`, as it is more of a developer shortcut, since it sets the same
+set of environment variables in all components of an app. This makes each
+sandbox bigger than necessary and doesn't emphasize the component model's
+strengths. See the [Spin Application Variables
+section](#spin-application-variables) for a discussion on how variables will be
+exposed to application components.
+
+#### Spin Application Variables
+
+Instead of configuring their applications with environment variables, SpinKube
+users should use [Spin application
+variables](https://developer.fermyon.com/spin/v2/variables#application-variables).
+
+These variables are declared in a Spin application manifest (`spin.toml`) and
+their values can be configured in the [`SpinApp`
+CRD](https://www.spinkube.dev/docs/spin-operator/reference/spin-app/#spinappspecvariablesindex)
+if using the environment variable provider. Application variables can also be
+dynamically updated using [Azure Key
+Vault](https://developer.fermyon.com/spin/v2/dynamic-configuration#azure-key-vault-application-variable-provider)
+or
+[Vault](https://developer.fermyon.com/spin/v2/dynamic-configuration#azure-key-vault-application-variable-provider).
+
+However, requiring that all variables be set by an administrator is likely too
+limiting for Kubernetes contexts where there are some cases in which the
+scheduler or webhooks inject environment variables into containers that
+components wish to access. For example, by default Kubernetes injects [cluster
+information environment
+variables](https://kubernetes.io/docs/concepts/containers/container-environment/#cluster-information)
+such as `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` into each
+container. Many Kubernetes libraries expect these to be available and would be
+incompatible otherwise. However, the shim should not simply inject all container
+environment variables into every component's environment. Instead, the shim will
+make all container environment variables available as Spin application variables
+by resetting them in the environment with the `SPIN_VARIABLE` prefix. This means
```

Comment:
I am not sure the Spin shim should do this for all variables. Could it parse the `spin.toml` and only elevate the environment variables that are explicitly requested?

What happens if a user sets a `SPIN_VARIABLE` on the containers env spec?  you might end up with `SPIN_VARIABLE_SPIN_VARIABLE_FOO`?

Reply:
I didn't follow this specific section in the SKIP. Are you saying that the shim will take every container variable (e.g. `KUBERNETES_SERVICE_HOST`), and append `SPIN_VARIABLE` prefix, and make them available to the every component? 

Taking it literally, that means the Spin components are seeing variables like `SPIN_VARIABLE_KUBERNETES_SERVICE_HOST`. How does this map to `kubernetes_service_address` in the Spin manifest? Will Spin runtime strips the prefix when it parses these `SPIN_VARIABLE` prefixed variables?

Reply:
> Are you saying that the shim will take every container variable (e.g. KUBERNETES_SERVICE_HOST), and append SPIN_VARIABLE prefix,

Yes

>  and make them available to the every component?

No. The Spin.toml defines what variables are available to which components. The prefix makes the variables available to the provider which will only provide the variable to components with explicit access to certain variables declared in the spin.toml

> Will Spin runtime strips the prefix when it parses these SPIN_VARIABLE prefixed variables?

Yes, this is how the environment variable variable provider works today

Reply:
will all the variables also be available without the prefix so the runtime can access them?  Is there any downside to parsing the spin.toml and only elevate the environment variables that are explicitly requested?

Reply:
Yes all envs will also be available as originally set -- we are creating a prefixed duplicate. See the example implementation: https://github.com/spinkube/containerd-shim-spin/pull/149

Referencing the spin.toml is a more precise option. I can switch the draft implementation to that approach

Reply:
sounds good, do we want to guard or warn against `SPIN_VARIABLE_SPIN_VARIABLE_FOO` situation?

Reply:
updated the draft PR. In that situation, is the variable in the spin.toml named "spin-variable-foo"? I think the implementation path i took should handle that scenario and still create a new env var called `SPIN_VARIABLE_SPIN_VARIABLE_FOO`. 

If the variable name is "foo", there is never a case in which `SPIN_VARIABLE_SPIN_VARIABLE_FOO` would be set in the current implementation. However, if both `FOO` and `SPIN_VARIABLE_FOO` are already set, `SPIN_VARIABLE_FOO` will be overwritten with `FOO`'s value in the current implementation. We can add another check if we think `SPIN_VARIABLE_FOO` should take precedence, but that does mean another env var check for every var

Reply:
oh, I guess it is no longer an issue since we are looking it up in the spin toml now.  I think we are allset 👍